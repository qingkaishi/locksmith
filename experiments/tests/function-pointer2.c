#include <pthread.h>
struct foo {
   void (*f)(void) ;
};

void f(void) { }

struct foo x = 0;

void g(struct foo *ht) {
  ht->f();
}

int i = 0;

void *thread_main(void *arg ) 
{
  g(&x);
  i = 0;//race
  return 0;
}
int main(int argc , char **argv ) 
{
  pthread_t t1;
  x.f = f;
  pthread_create(&t1, 0, & thread_main, 0);
  i = 0;//race
  return 0;
}
