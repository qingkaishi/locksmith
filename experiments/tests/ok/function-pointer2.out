************************* STARTING *************************

shared: &i:function-pointer2.c:14

Warning: Possible data race: &i:function-pointer2.c:14 is not protected!
 references:
  dereference of &i:function-pointer2.c:14 at function-pointer2.c:19
    &i:function-pointer2.c:14
  locks acquired:
    <empty>
  in: main at function-pointer2.c:22 

  dereference of &i:function-pointer2.c:14 at function-pointer2.c:27
    &i:function-pointer2.c:14
  locks acquired:
    <empty>
  in: main at function-pointer2.c:22 


*************************** DONE ***************************
